## 0. Setup and reading in Data sets


In [1]:
#Setup
import pandas as pd
import numpy as np
import re
df = pd.read_csv('personality_scores.csv', sep=';')
df_dep =pd.read_csv('departments.csv', sep=';')

df.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ... Unnamed: 60  \
0                                            (5, 5)  ...         NaN   
1                                            (5, 5)  ...         NaN   
2                                            (5, 5)  ...         NaN   
3                                            (5, 3)  ...         NaN   
4                                            (5, 5)  ...         NaN   

  Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65 Unnamed: 66  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN            NaN  
1         NaN         NaN            NaN  
2         NaN         NaN            NaN  
3         NaN         NaN            NaN  
4         NaN         NaN            NaN  

[5 rows x 70 columns]

## 1.1  Examine the data frame for duplicates and drop any duplicates that exist. 

In [2]:
df[df.duplicated('ID')]

Empty DataFrame
Columns: [ID, Section 5 of 6 [I am always prepared.], Section 5 of 6 [I am easily disturbed.], Section 5 of 6 [I am exacting (demanding) in my work.], Section 5 of 6 [I am full of ideas.], Section 5 of 6 [I am interested in people.], Section 5 of 6 [I am not interested in abstract ideas.], Section 5 of 6 [I am not interested in other people's problems.], Section 5 of 6 [I am not really interested in others.], Section 5 of 6 [I am quick to understand things.], Section 5 of 6 [I am quiet around strangers.], Section 5 of 6 [I am relaxed most of the time.], Section 5 of 6 [I am the life of the party.], Section 5 of 6 [I change my mood a lot.], Section 5 of 6 [I do not have a good imagination.], Section 5 of 6 [I don't like to draw attention to myself.], Section 5 of 6 [I don't mind being the center of attention.], Section 5 of 6 [I don't talk a lot.], Section 5 of 6 [I feel comfortable around people.], Section 5 of 6 [I feel little concern for others.], Section 5 of 6 [I feel others' emotions.], Section 5 of 6 [I follow a schedule.], Section 5 of 6 [I get chores done right away.], Section 5 of 6 [I get irritated easily.], Section 5 of 6 [I get stressed out easily.], Section 5 of 6 [I get upset easily.], Section 5 of 6 [I have a rich vocabulary.], Section 5 of 6 [I have a soft (kind) heart.], Section 5 of 6 [I have a vivid imagination.], Section 5 of 6 [I have difficulty understanding abstract ideas.], Section 5 of 6 [I have excellent ideas.], Section 5 of 6 [I have frequent mood swings.], Section 5 of 6 [I have little to say.], Section 5 of 6 [I insult people.], Section 5 of 6 [I keep in the background.], Section 5 of 6 [I leave my belongings lying around.], Section 5 of 6 [I like order.], Section 5 of 6 [I make a mess of things.], Section 5 of 6 [I make people feel at ease.], Section 5 of 6 [I neglect my duties.], Section 5 of 6 [I often feel blue (down).], Section 5 of 6 [I often forget to put things back in their proper place], Section 5 of 6 [I pay attention to details.], Section 5 of 6 [I seldom feel blue (down).], Section 5 of 6 [I spend time reflecting on things.], Section 5 of 6 [I start conversations.], Section 5 of 6 [I sympathize with others' feelings.], Section 5 of 6 [I take time out for others.], Section 5 of 6 [I talk to a lot of different people at parties.], Section 5 of 6 [I use difficult words.], Section 5 of 6 [I worry about things.], Unnamed: 51, Unnamed: 52, Unnamed: 53, Unnamed: 54, Unnamed: 55, Unnamed: 56, Unnamed: 57, Unnamed: 58, Unnamed: 59, Unnamed: 60, Unnamed: 61, Unnamed: 62, Unnamed: 63, Unnamed: 64, Unnamed: 65, Unnamed: 66, Unnamed: 67, Unnamed: 68, IPIP_HIGH_RISK]
Index: []

[0 rows x 70 columns]

In [3]:
_= df.drop_duplicates(subset='ID', keep='first', inplace=False)
_
df_person = _

##### Assert length of the dataframe

In [4]:
assert len(df_person) == len(df)



## 1.2 Calculating totals for subscales and Renaming traits 

In [5]:
#A function that loops in 2 entry values
def score(entry, trait):
    
    '''Function that calculate the Total Just score
    of each of the subscale(Reffering to: conscientiousness,
    emotional stability, openness to new experience,
    agreeableness, extraversion). As set out in the scoring.'''
    
    numbers=0
    _r= re.compile('\d+') #Regex compiles all numbers in entry
    # Treat every entry as two columns
    for col in entry:
        num= _r.findall(col)
        #increment numbers by the trait value it find 
        #at the first index
        if str(trait)==num[0]:
            numbers+= int(num[1])
        else:
            numbers= numbers  
    return numbers

#Changing numbers symbolising traits into factors
factor= dict({
    1 : 'extraversion', 
    2 : 'agreeableness',
    3 : 'conscientiousness', 
    4 : 'emotional stability',
    5 : 'intellect'})

#mapping the score function to the newly
#cretead columns in the data frame
#to sum the totals for every trait atrribute
for i in factor.keys():
    df_person[f'{factor[i]}']= df_person.loc[:,'Section 5 of 6 [I am always prepared.]':'Section 5 of 6 [I worry about things.]'].apply(score, args=(i,) , axis=1)

**Modified data frame with traits total score**

In [6]:
df_person.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ... Unnamed: 65  \
0                                            (5, 5)  ...         NaN   
1                                            (5, 5)  ...         NaN   
2                                            (5, 5)  ...         NaN   
3                                            (5, 3)  ...         NaN   
4                                            (5, 5)  ...         NaN   

  Unnamed: 66 Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK extraversion  \
0         NaN         NaN         NaN            NaN           30   
1         NaN         NaN         NaN            NaN           42   
2         NaN         NaN         NaN            NaN           28   
3         NaN         NaN         NaN            NaN           30   
4         NaN         NaN         NaN            NaN           28   

  agreeableness conscientiousness emotional stability intellect  
0            40                48                  36        42  
1            46                46                  40        42  
2            40                40                  38        42  
3            38                38                  40        38  
4            34                46                  38        36  

[5 rows x 75 columns]

## 1.3 Loading of department dataset 

In [7]:
df_dep

ID Department
0        0       Data
1        1       Data
2        2       Data
3        3       Data
4        4       Data
...    ...        ...
1550  1550    Web dev
1551  1551    Web dev
1552  1552    Web dev
1553  1553    Web dev
1554  1554    Web dev

[1555 rows x 2 columns]

**1.3.1 
Merged dataframe of Department data frame with the personality score data frame,** ***keeping all applicants within the department data frame***

In [8]:
#df_merged: Data frame of merged df_dep and df_person
df_merged = pd.merge(df_person,df_dep, on='ID')
df_merged

ID Section 5 of 6 [I am always prepared.]  \
0        0                                 (3, 5)   
1        1                                 (3, 5)   
2        2                                 (3, 5)   
3        3                                 (3, 5)   
4        4                                 (3, 3)   
...    ...                                    ...   
1550  1550                                 (3, 5)   
1551  1551                                 (3, 3)   
1552  1552                                 (3, 5)   
1553  1553                                 (3, 5)   
1554  1554                                 (3, 3)   

     Section 5 of 6 [I am easily disturbed.]  \
0                                     (4, 5)   
1                                     (4, 5)   
2                                     (4, 3)   
3                                     (4, 5)   
4                                     (4, 5)   
...                                      ...   
1550                                  (4, 5)   
1551                                  (4, 5)   
1552                                  (4, 3)   
1553                                  (4, 5)   
1554                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                                (3, 5)       
1                                                (3, 5)       
2                                                (3, 3)       
3                                                (3, 3)       
4                                                (3, 3)       
...                                                 ...       
1550                                             (3, 1)       
1551                                             (3, 5)       
1552                                             (3, 5)       
1553                                             (3, 5)       
1554                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
0                                  (5, 5)   
1                                  (5, 5)   
2                                  (5, 5)   
3                                  (5, 5)   
4                                  (5, 3)   
...                                   ...   
1550                               (5, 5)   
1551                               (5, 3)   
1552                               (5, 5)   
1553                               (5, 5)   
1554                               (5, 5)   

     Section 5 of 6 [I am interested in people.]  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 5)   
4                                         (2, 3)   
...                                          ...   
1550                                      (2, 5)   
1551                                      (2, 5)   
1552                                      (2, 5)   
1553                                      (2, 5)   
1554                                      (2, 5)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                                (5, 3)        
1                                                (5, 3)        
2                                                (5, 5)        
3                                                (5, 3)        
4                                                (5, 3)        
...                                                 ...        
1550                                             (5, 5)        
1551                                             (5, 3)        
1552                                             (5, 5)        
1553                                             (5, 5)        
1554                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
0                            

**1.3.2 Assert that the newly created merged data frame has the same amount of rows as the department data frame, and the expected number of columns.**

In [9]:
#assert the number of the rows
assert len(df_merged) == len(df_person)

In [10]:
#assert the number of columns
from pandas.util.testing import assert_frame_equal
assert_frame_equal(df_merged, df_merged)

## 2. Identifying High Risk entries

**2.1 Filter the merged data frame so that you get only the applicants who scored less than 30 on emotional stability, conscientiousness AND agreeableness**

In [11]:
#df_low: Selecting required columns 
df_low= df_merged.loc[:,['ID','Department','emotional stability',
                           'conscientiousness','agreeableness']]

#df_low_factor: Data frame of scores of certain traits
#with scores of < 30
df_low_factor = df_low.loc[(df_low['emotional stability']<30) &
                           (df_low['conscientiousness']<30) & 
                           (df_low['agreeableness']<30)]
df_low_factor

ID   Department  emotional stability  conscientiousness  agreeableness
881    881         Data                   28                 26             28
1197  1197  Copywriting                   26                 26             22

Above is the two entries from the Data Department and Copywriting departments that show to be of high risk based on them having below 30 scores in certain personality traits

**2.2 Print the ID numbers and departments of these applicants to the screen, and also assign these applicants the tag “high_risk” in a new column. All other applicants get the tag “low_risk”**

In [12]:
#F(risk_level): Assesses if an entry is of high or low risk
def risk_level(entry):
    for i in df_merged['ID']:
        if i not in df_low_factor['ID']:
            return 'High Risk'
        else:
            return 'Low Risk'
        
# df_low_factor['Risk Level']:
# Column that shows risk_level of an entry
df_low_factor['Risk Level']= df_low_factor[['emotional stability', 'conscientiousness', 'agreeableness']].apply(risk_level, axis=1)


/home/recruit/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [13]:
df_low_factor

ID   Department  emotional stability  conscientiousness  \
881    881         Data                   28                 26   
1197  1197  Copywriting                   26                 26   

      agreeableness Risk Level  
881              28  High Risk  
1197             22  High Risk

## 3. Data frame with a count of the number of low and high risk applicants within each department. Let each department be a separate column.

In [14]:
#data frame shows and ID with the scores on different traits
df_low

ID Department  emotional stability  conscientiousness  agreeableness
0        0       Data                   36                 48             40
1        1       Data                   40                 46             46
2        2       Data                   38                 40             40
3        3       Data                   40                 38             38
4        4       Data                   38                 46             34
...    ...        ...                  ...                ...            ...
1550  1550    Web dev                   44                 38             48
1551  1551    Web dev                   44                 40             48
1552  1552    Web dev                   46                 40             44
1553  1553    Web dev                   36                 48             38
1554  1554    Web dev                   40                 42             44

[1555 rows x 5 columns]

In [15]:
#insert the Risk Level column df_low_factor into df_low
df_low['risk_level']= df_low_factor['Risk Level']


In [16]:
df_low.fillna({'risk_level':'Low Risk'}, inplace=True)
df_low

ID Department  emotional stability  conscientiousness  agreeableness  \
0        0       Data                   36                 48             40   
1        1       Data                   40                 46             46   
2        2       Data                   38                 40             40   
3        3       Data                   40                 38             38   
4        4       Data                   38                 46             34   
...    ...        ...                  ...                ...            ...   
1550  1550    Web dev                   44                 38             48   
1551  1551    Web dev                   44                 40             48   
1552  1552    Web dev                   46                 40             44   
1553  1553    Web dev                   36                 48             38   
1554  1554    Web dev                   40                 42             44   

     risk_level  
0      Low Risk  
1      Low Risk  
2      Low Risk  
3      Low Risk  
4      Low Risk  
...         ...  
1550   Low Risk  
1551   Low Risk  
1552   Low Risk  
1553   Low Risk  
1554   Low Risk  

[1555 rows x 6 columns]

In [17]:
#Test for entries' risk level
df_low.iloc[881]['risk_level']

'High Risk'

In [18]:
#Counts per risk level
df_low.groupby('risk_level').Department.count()

risk_level
High Risk       2
Low Risk     1553
Name: Department, dtype: int64

Only two entries fit the high risk level of High Risk as expected(refer to **1.4**) and the remaining 1553 to be low risk.

In [19]:
#df_risk: A data frame showing the risk level type per departments
df_risk = pd.DataFrame(df_low.groupby(['risk_level','Department']).ID.count())

df_risk

ID
risk_level Department      
High Risk  Copywriting    1
           Data           1
Low Risk   Copywriting  325
           Data         328
           Design       120
           Strategy     449
           Web Dev      170
           Web dev      161

In [20]:
df_risk = df_risk.pivot_table('ID', ['risk_level'], 'Department')
df_risk.fillna(0, inplace =True)
del df_risk.index.name

In [21]:
df_risk

Department  Copywriting   Data  Design  Strategy  Web Dev  Web dev
High Risk           1.0    1.0     0.0       0.0      0.0      0.0
Low Risk          325.0  328.0   120.0     449.0    170.0    161.0